contracts/STYLE_GUIDE.md

```diff
@@ -0,0 +1,347 @@
+- 🤔Why not follow the Solidity recommendation of grouping by visibility? Visibility is clearly defined next to the method signature, making it trivial to check. However, searching can be deceiving because of inherited methods. Given this inconsistency in grouping, we find it easier to read and more consistent to organize code around functionality. Additionally, we recommend testing the public interface for any Solidity contract to ensure it only exposes expected methods.
+- Follow the [Solidity folder structure CLIP](https://github.com/smartcontractkit/CLIPs/tree/main/clips/2023-04-13-solidity-folder-structure)
+
+### Comments
+
+- Comments should be in the `//` (default) or `///` (natspec) format, not the `/*  */` format.
+- Comments should follow [NatSpec](https://docs.soliditylang.org/en/latest/natspec-format.html)
+- Besides comment above functions/structs, comments should live everywhere a reader might be confused. Don’t overestimate the reader of your contract, expect confusion in many places and document accordingly. This will help massively during audits and onboarding new team members.
+- Headers should be used to group functionality, the following header style and length is recommended.
+  - Don’t use headers for a single function, or to say “getters”. Group by functionality e.g. the `Tokens and pools` , or `fees` logic within the CCIP OnRamp.
+
+```solidity
+  // ================================================================
+  // │                      Tokens and pools                        │
+  // ================================================================
+
+....
+
+  // ================================================================
+  // │                             Fees                             │
+  // ================================================================
+```
+
+### Imports
+
+- Imports should always be explicit
+- Imports should follow the following format:
+
+```solidity
+import {IInterface} from "../interfaces/IInterface.sol";
+
+import {AnythingElse} from "../code/AnythingElse.sol";
+
+import {ThirdPartyCode} from "../../vendor/ThirdPartyCode.sol";
+```
+
+- An example would be
+
+```solidity
+import {ITypeAndVersion} from "../../shared/interfaces/ITypeAndVersion.sol";
+import {IPool} from "../interfaces/pools/IPool.sol";
+
+import {AggregateRateLimiter} from "../AggregateRateLimiter.sol";
+import {Client} from "../libraries/Client.sol";
+
+import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.0/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "../../vendor/openzeppelin-solidity/v4.8.0/contracts/token/ERC20/IERC20.sol";
+```
+
+- Sorting imports alphabetically within the sections is nice, but since we don’t have any tooling to automatically do this it is not required.
```

Comment:
There's definitely an eslint rule for this. I'd say that this shouldn't be mentioned here, as otherwise someone who prefers this might make style comments, and these can be annoying.

Maybe it would be worthwhile to add a general rule of thumb that I use: "Anything to do with style should be a formatting plugin. If you care enough about a particular style, add or write that plugin, so that others wouldn't need to manually do this."

Reply:
I've split the doc into a Guidelines and a Rules section. All entries in the Rules section have a solhint rule